### ISSUE TITLE
Wrong value/infinite loop in for-loop when using iterators with tail calls

### ISSUE THREAD

**Issue by xconstruct** [yuin/gopher-lua#82]
Oh well, sorry, I ran into a second issue :)

When I use a coroutine together with a for-loop, the second value is wrong - the first value is duplicated. If I call the coroutine manually, everything is fine.

Example:
```lua
local cr = function()
        return coroutine.wrap(function()
                coroutine.yield(1, "a")
                coroutine.yield(2, "b")
        end)
end

for k, v in cr() do
        print("for:", k, v)
end

print("------")

local f = cr()
print("call:", f())
print("call:", f())
```

Output:
```
for:    1       1
for:    2       2
------
call:   1       a
call:   2       b
```

Thanks again!

**Comment by xconstruct** [comment]
I tried to find a workaround for my iterators, but I noticed it does not only affect coroutines. I think it is a problem with tail calls in the iterator function, maybe the stack gets confused.

This example results in an endless loop:
```lua
local get_val = function(i)
        if i == 1 then return i, "a" end
        if i == 2 then return i, "b" end
        return nil
end

local my_next = function(invariant, i)
        i = (i or 0) + 1
        print("next:", i)
        return get_val(i)
end

for k, v in my_next do
        print("for:", k, v)
end
```

Output:
```
next:   1
for:    1       1
for:    1       1
for:    1       1
... and so on
```

It looks like the for-loop tries to call the inner function `get_val` and not the iterator `my_next`. It does not happen if we inline the `get_val`:

```lua
local my_next = function(invariant, i)
        i = (i or 0) + 1
        print("next:", i)
        if i == 1 then return i, "a" end
        if i == 2 then return i, "b" end
        return nil
end

for k, v in my_next do
        print("for:", k, v)
end
```

Works correctly:
```
next:   1
for:    1       a
next:   2
for:    2       b
next:   3
```

**Comment by yuin** [comment]
Thank you again, @xconstruct . 
I've fixed the bug in dd0fd531098d425ac1576765d9cca3274f8dd8d6 thanks to your detailed report :thumbsup:
